src/core/lv2/Lv2Worker.cpp

```diff
@@ -0,0 +1,210 @@
+	// for jack, they use 4 * jack_port_type_get_buffer_size (..., JACK_DEFAULT_MIDI_TYPE)
+	// (possible extension for AudioDevice)
+}
+
+
+
+
+Lv2Worker::Lv2Worker(LV2_Handle handle,
+	const LV2_Worker_Interface* iface,
+	Semaphore* common_work_lock,
+	bool threaded) :
+	m_iface(iface),
+	m_threaded(threaded),
+	m_handle(handle),
+	m_response(bufferSize()),
+	m_requests(bufferSize()),
+	m_responses(bufferSize()),
+	m_requestsReader(m_requests),
+	m_responsesReader(m_responses),
+	m_sem(0),
+	m_workLock(common_work_lock)
+{
+	assert(iface);
+	m_scheduleFeature.handle = static_cast<LV2_Worker_Schedule_Handle>(this);
+	m_scheduleFeature.schedule_work = [](LV2_Worker_Schedule_Handle handle,
+		uint32_t size, const void* data) -> LV2_Worker_Status
+		{
+			Lv2Worker* worker = static_cast<Lv2Worker*>(handle);
+			return worker->scheduleWork(size, data);
+		};
+
+	if (threaded) { m_thread = std::thread(&Lv2Worker::workerFunc, this); }
+
+	m_requests.mlock();
+	m_responses.mlock();
+}
+
+
+
+
+Lv2Worker::~Lv2Worker()
+{
+	m_exit = true;
+	if(m_threaded) {
+		m_sem.post();
+		m_thread.join();
+	}
+}
+
+
+
+
+LV2_Worker_Status Lv2Worker::respond(uint32_t size, const void* data)
+{
+	if(!size) { return LV2_WORKER_ERR_UNKNOWN; }
+
+	LV2_Worker_Status rval = LV2_WORKER_SUCCESS;
+	if(m_threaded)
+	{
+		if(m_responses.free() < sizeof(size) + size)
+		{
+			rval = LV2_WORKER_ERR_NO_SPACE;
+		}
+		else
+		{
+			m_responses.write((const char*)&size, sizeof(size));
+			m_responses.write((const char*)data, size);
+		}
+	}
+	else
+	{
+		m_iface->work_response(m_handle, size, data);
+	}
+	return rval;
+}
+
+
+
+
+void Lv2Worker::workerFunc()
+{
+	std::vector<char> buf;
+	uint32_t size;
+	while (true) {
+		m_sem.wait();
+		if (m_exit) break;
+		const std::size_t readSpace = m_requestsReader.read_space();
+		if (readSpace <= sizeof(size)) continue; // (should not happen)
+
+		m_requestsReader.read(sizeof(size)).copy((char*)&size, sizeof(size));
+		assert(size <= readSpace - sizeof(size));
+		if(size > buf.size())
+			buf.resize(size);
+		m_requestsReader.read(size).copy(buf.data(), size);
+
+		m_workLock->wait();
+		m_iface->work(m_handle, staticWorkerRespond, this, size, buf.data());
+		m_workLock->post();
+	}
+}
+
+
+
+
+LV2_Worker_Status Lv2Worker::scheduleWork(uint32_t size, const void *data)
+{
+	if (!size) { return LV2_WORKER_ERR_UNKNOWN; }
+
+	LV2_Worker_Status rval = LV2_WORKER_SUCCESS;
+	if (m_threaded)
+	{
+		if(m_requests.free() < sizeof(size) + size)
+		{
+			rval = LV2_WORKER_ERR_NO_SPACE;
+		}
+		else
+		{
+			// Schedule a request to be executed by the worker thread
+			m_requests.write((const char*)&size, sizeof(size));
+			m_requests.write((const char*)data, size);
+			m_sem.post();
+		}
+	}
+	else
+	{
+		// Execute work immediately in this thread
+		m_workLock->wait();
+		m_iface->work(m_handle, staticWorkerRespond, this, size, data);
+		m_workLock->post();
+	}
+
+	return rval;
+}
+
+
+
+
+void Lv2Worker::emitResponses()
+{
+	if (!m_exit) {
```

Comment:
What's the purpose of this condition? The documentation for `LV2_Worker_Respond_Function` states "the host MUST guarantee that [the data] will be eventually passed to `work_response()` if this function returns `LV2_WORKER_SUCCESS`", so doing so conditionally seems to violate the requirements. It also doesn't seem relevant to this function - it runs on the real-time thread, not the worker thread.

Reply:
Agreed, this variable is only for the worker thread. I guess I wanted to make this symmetric. Anyways, will remove this `if`.